internal/cmd/update.go

```diff
@@ -0,0 +1,309 @@
+	//   * coder assets root (env CODER_ASSETS_ROOT)
+
+	currentBinaryStat, err := u.fs.Stat(u.executablePath)
+	if err != nil {
+		return clog.Fatal("preflight: cannot stat current binary", clog.Causef("%s", err))
+	}
+
+	if currentBinaryStat.Mode().Perm()&0222 == 0 {
+		return clog.Fatal("preflight: missing write permission on current binary")
+	}
+
+	var coderURL *url.URL
+	if coderURLString == "" {
+		coderURL, err = getCoderConfigURL()
+		if err != nil {
+			return clog.Fatal(
+				"Unable to automatically determine coder URL",
+				clog.Causef(err.Error()),
+				clog.BlankLine,
+				clog.Tipf("use --coder <url> to specify coder URL"),
+			)
+		}
+	} else {
+		coderURL, err = url.Parse(coderURLString)
+		if err != nil {
+			return clog.Fatal("invalid coder URL", err.Error())
+		}
+	}
+
+	desiredVersion, err := getAPIVersionUnauthed(u.httpClient, *coderURL)
+	if err != nil {
+		return clog.Fatal("fetch api version", clog.Causef(err.Error()))
+	}
+
+	clog.LogInfo(fmt.Sprintf("Coder instance at %q reports version %s", coderURL.String(), desiredVersion.String()))
+	clog.LogInfo(fmt.Sprintf("Current version of coder-cli is %s", version.Version))
+
+	if currentVersion, err := semver.Make(u.versionF()); err == nil {
+		if desiredVersion.Compare(currentVersion) == 0 {
+			clog.LogInfo("Up to date!")
+			return nil
+		}
+	}
+
+	if !force {
+		label := fmt.Sprintf("Update coder-cli to version %s", desiredVersion.FinalizeVersion())
+		if _, err := u.confirmF(label); err != nil {
+			return clog.Fatal("failed to confirm update", clog.Tipf(`use "--force" to update without confirmation`))
+		}
+	}
+
+	downloadURL := makeDownloadURL(desiredVersion.FinalizeVersion(), runtime.GOOS, runtime.GOARCH)
+
+	var downloadBuf bytes.Buffer
+	memWriter := bufio.NewWriter(&downloadBuf)
+
+	clog.LogInfo("fetching coder-cli from GitHub releases", downloadURL)
+	resp, err := u.httpClient.Get(downloadURL)
+	if err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to fetch URL %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return clog.Fatal("failed to fetch release", clog.Causef("URL %s returned status code %d", downloadURL, resp.StatusCode))
+	}
+
+	if _, err := io.Copy(memWriter, resp.Body); err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to download %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	_ = resp.Body.Close()
+
+	if err := memWriter.Flush(); err != nil {
+		return clog.Fatal(fmt.Sprintf("failed to save %s", downloadURL), clog.Causef(err.Error()))
+	}
+
+	// TODO: validate the checksum of the downloaded file. GitHub does not currently provide this information
+	// and we do not generate them yet.
+	updatedBinary, err := extractFromArchive("coder", downloadBuf.Bytes())
+	if err != nil {
+		return clog.Fatal("failed to extract coder binary from archive", clog.Causef(err.Error()))
+	}
+
+	// We assume the binary is named coder and write it to coder.new
+	updatedCoderBinaryPath := u.executablePath + ".new"
```

Comment:
An implicit assumption of this code is that users either:

1. Only use a single coder installation; or
2. Use the same version of coder across multiple deployments (e.g. dev/stage/test)

I don't think we should make this assumption in general, and would propose a scheme where coder-cli has a bootstrap portion which looks for existing commands, updates if needed, and then exec's that. For example:

* `coder` command is a small shim that does this release check
* check the server version (suppose for this example it's 1.20.0)
* if the versions don't match, download the corresponding version, and save as `~/.coder/bin/coder-1.20.0.exe`)
* exec that command and pass along arguments

This way, a user can transparently switch between different installations of Coder, and always have an up-to-date version of the tool. And we don't have to worry about long-term compatibility guarantees between client and server (for now)

Reply:
One other thing to consider -- the `oc` command (for OpenShift) and busybox have magic behavior where the command that runs depends on `argv[0]`, which means that we could reuse the `coder` CLI command for the bootstrap logic as well as the command itself. That is, if the program is run as `coder`, then do the bootstrapping I described above, and if the program is run as `coder-1.21.0`, then run the coder CLI commands.

This means that we don't have to create a new build or change the way things are built or delivered, but we benefit from the above logic

For the way oc does it, check here: https://github.com/openshift/oc/blob/25b5251c48ea8e3d1294e6b9b1c05bebb29de46f/cmd/oc/oc.go#L91-L93

Reply:
Having `coder-cli` be able to manage multiple versions of itself would be a major usability win.

However, I think it's a bit out of scope of this issue, so I created a separate story to follow up on it.